libyul/optimiser/MemoryEscalator.cpp

```diff
@@ -0,0 +1,139 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <libyul/optimiser/MemoryEscalator.h>
+#include <libyul/optimiser/CallGraphGenerator.h>
+#include <libyul/optimiser/FunctionCallFinder.h>
+#include <libyul/optimiser/NameDispenser.h>
+#include <libyul/optimiser/StackToMemoryMover.h>
+#include <libyul/backends/evm/EVMDialect.h>
+#include <libyul/AsmData.h>
+#include <libyul/CompilabilityChecker.h>
+#include <libyul/Dialect.h>
+#include <libyul/Exceptions.h>
+#include <libyul/Object.h>
+#include <libyul/Utilities.h>
+#include <libsolutil/Algorithms.h>
+#include <libsolutil/CommonData.h>
+
+using namespace std;
+using namespace solidity;
+using namespace solidity::yul;
+
+MemoryOffsetAllocator::MemoryOffsetAllocator(
+	map<YulString, FunctionStackErrorInfo> const& _functionStackErrorInfo,
+	set<YulString> const& _functionsInCycle,
+	map<YulString, set<YulString>> const& _callGraph
+): m_functionStackErrorInfo(_functionStackErrorInfo), m_functionsInCycle(_functionsInCycle), m_callGraph(_callGraph)
+{
+}
+
+bool MemoryOffsetAllocator::run() { return run({}); }
+uint64_t MemoryOffsetAllocator::requiredSlots() const {	return m_nextAvailableSlot.at({}); }
+
+bool MemoryOffsetAllocator::run(YulString v)
+{
+	yulAssert(!m_nextAvailableSlot.count(v), "");
+	uint64_t& n = m_nextAvailableSlot[v];
+	if (m_callGraph.count(v))
+		for (auto child: m_callGraph.at(v))
+		{
+			if (!m_nextAvailableSlot.count(child))
+				if (!run(child))
+					return false;
+			n = std::max(n, m_nextAvailableSlot.at(child));
+		}
+
+	if (m_functionStackErrorInfo.count(v))
+	{
+		if (m_functionsInCycle.count(v))
+			return false;
```

Comment:
This looks like an exception to me.

Reply:
If you make this an exception, you can use the return value for more useful things, like returning `m_nextAvailableSlot[v]`.

Reply:
You mean an exception I catch in the memory escalator or an actual exception propagating to the outside?
So far I just had the memory escalator do nothing in cases like this, thinking that *something* might happen that still makes things go through in the end... And since I know what emscripten does with exceptions I'm always hesitant to throw them only to catch them :-). But yeah, it should probably be one here either way.

By the way: theoretically I think we could actually handle recursive calls using a Vyper-like calling convention...
If we move all its arguments and return values to memory, then we can store all escalated memory variables on *stack* before the recursive call and then, after the recursive call returns, restore the old memory variables from stack :-). Not that that's nice, but it should actually be a working option...